runner/src/examples.rs

```diff
@@ -0,0 +1,938 @@
+            &example.example.server,
+            opt.server_additional_args.clone(),
+            &self.out,
+            &opt.permissions_file,
+        );
+
+        if opt.build_client.client_variant == NO_CLIENTS {
+            Step::Single {
+                name: "run server".to_string(),
+                command: run_server,
+            }
+        } else {
+            Step::WithBackground {
+                name: "background server".to_string(),
+                background: run_server,
+                foreground: Box::new(run_clients),
+            }
+        }
+    }
+}
+
+impl ApplicationFunctions {
+    fn construct_application_build_steps(&self, example_name: &str) -> Vec<Step> {
+        vec![build_wasm_module(example_name, &self.target, example_name)]
+    }
+
+    fn construct_example_server_run_step(
+        &self,
+        example: &FunctionsExample,
+        run_clients: Step,
+    ) -> Step {
+        let opt = &example.options;
+        let run_server = run_functions_example_server(&example.example.server, &self);
+
+        if opt.build_client.client_variant == NO_CLIENTS {
+            Step::Single {
+                name: "run server".to_string(),
+                command: run_server,
+            }
+        } else {
+            Step::WithBackground {
+                name: "background server".to_string(),
+                background: run_server,
+                foreground: Box::new(run_clients),
+            }
+        }
+    }
+}
+
+trait OakExample {
+    fn get_backends(&self) -> &HashMap<String, Executable>;
+
+    fn get_build_client(&self) -> &BuildClient;
+
+    /// Constructs application build steps.
+    fn construct_application_build_steps(&self) -> Vec<Step>;
+
+    /// Constructs run step for the example server.
+    fn construct_example_server_run_step(&self, run_clients: Step) -> Step;
+
+    /// Constructs build steps for the backends.
+    fn construct_backend_build_steps(&self) -> Vec<Step> {
+        self.get_backends()
+            .iter()
+            .map(move |(name, backend)| Step::Single {
+                name: name.to_string(),
+                command: build(&backend.target, &self.get_build_client()),
+            })
+            .collect()
+    }
+
+    /// Recursively constructs run steps for the backends.
+    fn construct_backend_run_steps(&self, run_server_clients: Step) -> Step {
+        self.get_backends()
+            .iter()
+            // First iteration includes `run_server_clients` as a foreground step.
+            .fold(run_server_clients, |backend_steps, (name, backend)| {
+                Step::WithBackground {
+                    name: name.to_string(),
+                    // Each `backend` is included as background step.
+                    background: run(&backend, &self.get_build_client(), Vec::new()),
+                    foreground: Box::new(backend_steps),
+                }
+            })
+    }
+}
+pub struct ClassicExample<'a> {
+    example: &'a Example,
```

Comment:
Not sure why all the lifetimes and Box are needed, is it not possible to use objects owned by this struct?

Reply:
I tried. But the main problem was that I could not find a nice way to move the example objects from the vector in `run_examples` and pass them to the constructor of this class.